scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,824 @@
+                                                                                                                          BiFunction.class));
+                    logWithContext(taskid,
+                                   channel,
+                                   "Run conditionalReplaceAll(Else) on key '" + currentKey + "' result is: " + answer,
+                                   Level.DEBUG);
+
+                }
+            }
+            commitIfNeeded(channel);
+
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new ClosureEvaluationException("Exception when evaluating closure : " +
+                                                 StackTraceUtil.getStackTrace(e));
+        }
+    }
+
+    @Override
+    public <T extends Serializable> boolean waitUntil(String originator, TaskId taskid, String channel, String key,
+            String predicate) throws InvalidChannelException, CompilationException {
+        Channel<T> chosenChannel = getChannel(channel);
+        T value = chosenChannel.get(key);
+        try {
+            return evaluateClosure(predicate, BiPredicate.class).test(key, value);
+        } catch (CompilationException e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new ClosureEvaluationException("Exception when evaluating closure : " +
+                                                 StackTraceUtil.getStackTrace(e));
+        }
+    }
+
+    @Override
+    public <T extends Serializable> PredicateActionResult<T> waitUntilThen(String originator, TaskId taskid,
+            String channel, String key, String predicate, String thenRemappingFunction)
+            throws InvalidChannelException, CompilationException, IOException {
+        PredicateActionResult<T> answer;
+        try {
+            Channel<T> chosenChannel = getChannel(channel);
+            T value = chosenChannel.get(key);
+            if (evaluateClosure(predicate, BiPredicate.class).test(key, value)) {
+                answer = new PredicateActionResult<>(true,
+                                                     chosenChannel.compute(key,
+                                                                           (BiFunction<? super String, ? super T, ? extends T>) evaluateClosure(thenRemappingFunction,
+                                                                                                                                                BiFunction.class)));
+            } else {
+                throw new IllegalStateException("Predicate " + predicate + " is not met");
+            }
+            commitIfNeeded(channel);
+            logWithContext(taskid, channel, "Run waitUntil on key '" + key + "' result is: " + answer, Level.DEBUG);
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Throwable e) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)